src/newton_solver.cpp

```diff
@@ -105,144 +78,71 @@ void NewtonSolver::computeNewtonSensis(
+void NewtonSolver::prepareLinearSystem(
     Model& model, SimulationState const& state
 ) {
-    model.fJ(state.t, 0.0, state.x, state.dx, xdot_, Jtmp_);
-    Jtmp_.refresh();
-    auto status = SUNLinSolSetup_Dense(linsol_, Jtmp_);
-    if (status != SUNLS_SUCCESS)
-        throw NewtonFailure(status, "SUNLinSolSetup_Dense");
+    auto& J = linsol_->getMatrix();
+    if (J.matrix_id() == SUNMATRIX_SPARSE) {
+        model.fJSparse(state.t, 0.0, state.x, state.dx, xdot_, J);
+    } else {
+        model.fJ(state.t, 0.0, state.x, state.dx, xdot_, J);
+    }
+    J.refresh();
+    try {
+        linsol_->setup();
+    } catch (AmiException const& e) {
+        throw NewtonFailure(AMICI_SINGULAR_JACOBIAN, e.what());
+    }
 }
 
-void NewtonSolverDense::prepareLinearSystemB(
+void NewtonSolver::prepareLinearSystemB(
     Model& model, SimulationState const& state
 ) {
-    model.fJB(state.t, 0.0, state.x, state.dx, xB_, dxB_, xdot_, Jtmp_);
-    Jtmp_.refresh();
-    auto status = SUNLinSolSetup_Dense(linsol_, Jtmp_);
-    if (status != SUNLS_SUCCESS)
-        throw NewtonFailure(status, "SUNLinSolSetup_Dense");
+    auto& J = linsol_->getMatrix();
+    if (J.matrix_id() == SUNMATRIX_SPARSE) {
+        model.fJSparseB(state.t, 0.0, state.x, state.dx, xB_, dxB_, xdot_, J);
+    } else {
+        model.fJB(state.t, 0.0, state.x, state.dx, xB_, dxB_, xdot_, J);
+    }
+    J.refresh();
+    try {
+        linsol_->setup();
+    } catch (AmiException const& e) {
+        throw NewtonFailure(AMICI_SINGULAR_JACOBIAN, e.what());
+    }
 }
 
-void NewtonSolverDense::solveLinearSystem(AmiVector& rhs) {
-    auto status = SUNLinSolSolve_Dense(
-        linsol_, Jtmp_, rhs.getNVector(), rhs.getNVector(), 0.0
-    );
-    Jtmp_.refresh();
+void NewtonSolver::solveLinearSystem(AmiVector& rhs) {
     // last argument is tolerance and does not have any influence on result
-
+    auto status = linsol_->solve(rhs.getNVector(), rhs.getNVector(), 0.0);
     if (status != SUNLS_SUCCESS)
-        throw NewtonFailure(status, "SUNLinSolSolve_Dense");
+        throw NewtonFailure(status, "SUNLinSolSolve");
 }
 
-void NewtonSolverDense::reinitialize() {
-    /* dense solver does not need reinitialization */
+void NewtonSolver::reinitialize() {
+    // dense solver does not need reinitialization
+    if (auto s = dynamic_cast<SUNLinSolKLU*>(linsol_.get())) {
+        try {
+            s->reInit(s->getMatrix().capacity(), SUNKLU_REINIT_PARTIAL);
+        } catch (AmiException const&) {
+            throw NewtonFailure(AMICI_ERROR, "Failed to reinitialize KLU solver"
+            );
+        }
+    }
 };
 
-bool NewtonSolverDense::is_singular(Model& model, SimulationState const& state)
+bool NewtonSolver::is_singular(Model& model, SimulationState const& state)
     const {
+    if (auto s = dynamic_cast<SUNLinSolKLU const*>(linsol_.get())) {
+        return s->is_singular();
+    }
+
     // dense solver doesn't have any implementation for rcond/condest, so use
```

Comment:
what about other direct/implicit solvers?

Reply:
As above -- they haven't been supported before this PR, and they are still not supported yet. Although it would be pretty easy to extend that now....